packages/std/src/coins.rs

```diff
@@ -0,0 +1,392 @@
+    pub fn to_vec(&self) -> Vec<Coin> {
+        self.0
+            .iter()
+            .map(|(denom, amount)| Coin {
+                denom: denom.clone(),
+                amount: *amount,
+            })
+            .collect()
+    }
+
+    /// Conversion to Vec<Coin>, consuming the original object
+    pub fn into_vec(self) -> Vec<Coin> {
+        self.0
+            .into_iter()
+            .map(|(denom, amount)| Coin { denom, amount })
+            .collect()
+    }
+
+    /// Returns the number of different denoms in this collection.
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns `true` if this collection contains no coins.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Returns the denoms as a vector of strings.
+    /// The vector is guaranteed to not contain duplicates and sorted alphabetically.
+    pub fn denoms(&self) -> Vec<String> {
+        self.0.keys().cloned().collect()
+    }
+
+    /// Returns the amount of the given denom or zero if the denom is not present.
+    pub fn amount_of(&self, denom: &str) -> Uint128 {
+        self.0.get(denom).copied().unwrap_or_else(Uint128::zero)
+    }
+
+    /// Returns the amount of the given denom if and only if this collection contains only
+    /// the given denom. Otherwise `None` is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use cosmwasm_std::{Coin, Coins, coin};
+    ///
+    /// let coins: Coins = [coin(100, "uatom")].try_into().unwrap();
+    /// assert_eq!(coins.contains_only("uatom").unwrap().u128(), 100);
+    /// assert_eq!(coins.contains_only("uluna"), None);
+    /// ```
+    ///
+    /// ```rust
+    /// use cosmwasm_std::{Coin, Coins, coin};
+    ///
+    /// let coins: Coins = [coin(100, "uatom"), coin(200, "uusd")].try_into().unwrap();
+    /// assert_eq!(coins.contains_only("uatom"), None);
+    /// ```
+    pub fn contains_only(&self, denom: &str) -> Option<Uint128> {
+        if self.len() == 1 {
+            self.0.get(denom).copied()
+        } else {
+            None
+        }
+    }
+
+    /// Adds the given coin to this `Coins` instance.
+    /// Errors in case of overflow.
+    pub fn add(&mut self, coin: Coin) -> StdResult<()> {
+        if coin.amount.is_zero() {
+            return Ok(());
+        }
+
+        let amount = self.0.entry(coin.denom).or_insert_with(Uint128::zero);
+        *amount = amount.checked_add(coin.amount)?;
+        Ok(())
+    }
+
+    /// Subtracts the given coin from this `Coins` instance.
+    /// Errors in case of overflow or if the denom is not present.
+    pub fn sub(&mut self, coin: Coin) -> StdResult<()> {
+        match self.0.get_mut(&coin.denom) {
+            Some(v) => {
+                *v = v.checked_sub(coin.amount)?;
+                // make sure to remove zero coin
+                if v.is_zero() {
+                    self.0.remove(&coin.denom);
```

Comment:
This looks a bit scary. Can we remove an element from the map while holding a mutable reference to the value?

Reply:
Impressive